src/NServiceBus.Transport.Msmq/MessagePump.cs

```diff
@@ -187,13 +194,19 @@ Task ReceiveMessage()
                 catch (Exception ex)
                 {
                     Logger.Warn("MSMQ receive operation failed", ex);
-                    await messagePump.receiveCircuitBreaker.Failure(ex).ConfigureAwait(false);
+                    await receiveCircuitBreaker.Failure(ex).ConfigureAwait(false);
                 }
                 finally
                 {
-                    messagePump.concurrencyLimiter.Release();
+                    concurrencyLimiter.Release();
                 }
-            }, this);
+            },
+            null,
+            CancellationToken.None,  // CancellationToken.None is used here since cancelling the task before it can run can cause the concurrencyLimiter to not be released
+            TaskCreationOptions.DenyChildAttach,
+            TaskScheduler.Default)
+                .Unwrap();
```

Comment:
I'm confused by this. MSMQ is the only transport I've seen do this. Should all transports be doing it?

Reply:
Not sure, @danielmarbach do U remember if we had some reasons for optimizing just MSMQ?

Reply:
Pretty sure we have done this to make sure there is less closure allocation. It was an important optimization at the time because a single receive operation is offloaded to the threadpool for every message that is handled. So if we manage to avoid capturing less in the closure (for example by allowing to remove the Func allocation or in the best case make sure there is no closure at all anymore) this adds up very quickly considering the hundreds of messages per second the transport can handle.

Reply:
Here is some more context. https://github.com/Particular/NServiceBus/issues/3884 As you can see Gen0 can be sometimes halved. But it really depends on the code the is underneath and you need to look at the decompiled code before and after

Reply:
Assuming this stays in MSMQ, it needs a Big Nasty Comment so we don't have to refer back to https://github.com/Particular/NServiceBus/issues/3884.

@danielmarbach Are you saying that if we take the compiled result of any given transport and decompile it, we'll be able to see whether or not the "Go process the message and we don't care about the returned task" operation results in a closure, and then if it does, we should be using this Task.Run construction, and if not, leave it alone?

Reply:
Very crude example since it is already late

```
    class Program
    {
        static async Task Main(string[] args)
        {
            var state = new object();

            await Captures(state);

            await DoesNotCapture(state);
        }

        static async Task Captures(object state) {
            await Task.Run(async () => {
                await Console.Error.WriteLineAsync(state.GetHashCode().ToString());
            });
        }

        static async Task DoesNotCapture(object state) {
            await Task.Factory.StartNew(async s => {
                await Console.Error.WriteLineAsync(s.GetHashCode().ToString());
            }, state,
            CancellationToken.None,
            TaskCreationOptions.DenyChildAttach,
            TaskScheduler.Default)
            .Unwrap();
        }
    }
```

First one leads to 

```
            Program.\u003C\u003Ec__DisplayClass1_0 cDisplayClass10 = new Program.\u003C\u003Ec__DisplayClass1_0();
            cDisplayClass10.state = this.state;
            // ISSUE: method pointer
            awaiter = Task.Run(new Func<Task>((object) cDisplayClass10, __methodptr(\u003CCaptures\u003Eb__0))).GetAwaiter();
```

vs

```
            awaiter = Task.Factory.StartNew<Task>(Program.\u003C\u003Ec.\u003C\u003E9__2_0 ?? (Program.\u003C\u003Ec.\u003C\u003E9__2_0 = new Func<object, Task>((object) Program.\u003C\u003Ec.\u003C\u003E9, __methodptr(\u003CDoesNotCapture\u003Eb__2_0))), this.state, CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default).Unwrap().GetAwaiter();
```

To bring this into the per-message execution land the first one would create a display class and a Func allocation for every execution while the second one only statically allocates the Func ones and doesn't require a display class. So effectively two allocations per execution less. Though sometimes it is only possible to get rid of the Display class allocation and not the func allocation depending on the code you are trying to optimize.